lib/rouge/lexers/ada.rb

```diff
@@ -0,0 +1,146 @@
+      mimetypes 'text/x-ada'
+
+      title 'Ada'
+      desc 'The Ada 2012 programming language'
+
+      # Ada identifiers are Unicode with underscores only allowed as separators.
+      ID = /\b[[:alpha:]](?:\p{Pc}?[[:alnum:]])*\b/
+
+      # Numerals can also contain underscores.
+      NUM = /\d(_?\d)*/
+      XNUM = /\h(_?\h)*/
+      EXP = /(E[-+]?#{NUM})?/i
+
+      # Return a hash mapping lower-case identifiers to token classes.
+      def self.idents
+        @idents ||= Hash.new(Name).tap do |h|
+          %w(
+            abort abstract accept access aliased all array at begin body
+            case constant declare delay delta digits do else elsif end
+            exception exit for generic goto if in interface is limited
+            loop new null of others out overriding pragma private
+            protected raise range record renames requeue return reverse
+            select separate some synchronized tagged task terminate then
+            until use when while with
+          ).each {|w| h[w] = Keyword}
+
+          %w(abs and mod not or rem xor).each {|w| h[w] = Operator::Word}
+
+          %w(
+            entry function package procedure subtype type
+          ).each {|w| h[w] = Keyword::Declaration}
+
+          %w(
+            boolean character constraint_error duration float integer
+            natural positive long_float long_integer long_long_float
+            long_long_integer program_error short_float short_integer
+            short_short_integer storage_error string tasking_error
+            wide_character wide_string wide_wide_character
+            wide_wide_string
+          ).each {|w| h[w] = Name::Builtin}
+        end
+      end
+
+      state :whitespace do
+        rule %r{\s+}m, Text
+        rule %r{--.*$}, Comment::Single
+      end
+
+      state :dquote_string do
+        rule %r{[^"\n]+}, Literal::String::Double
+        rule %r{""}, Literal::String::Escape
+        rule %r{"}, Literal::String::Double, :pop!
+        rule %r{\n}, Error, :pop!
+      end
+
+      state :attr do
+        mixin :whitespace
+        rule ID, Name::Attribute, :pop!
+        rule %r{}, Text, :pop!
+      end
+
+      # Handle a dotted name immediately following a declaration keyword.
+      state :decl_name do
+        mixin :whitespace
+        rule %r{body\b}i, Keyword::Declaration  # package body Foo.Bar is...
+        rule %r{(#{ID})(\.)} do
+          groups Name::Namespace, Punctuation
+        end
+        # function "<=" (Left, Right: Type) is ...
+        rule %r{#{ID}|"(and|or|xor|/?=|<=?|>=?|\+|–|&\|/|mod|rem|\*?\*|abs|not)"},
+             Name::Function, :pop!
+        rule %r{}, Text, :pop!
+      end
+
+      # Handle a sequence of library unit names: with Ada.Foo, Ada.Bar;
+      state :libunit_name do
+        mixin :whitespace
+        rule ID do
+          token Name::Namespace
+          goto :libunit_sep
+        end
+        rule %r{}, Text, :pop!
+      end
+
+      state :libunit_sep do
```

Comment:
Is there a reason to have this state separate to `:libunit_name`? Is it just for syntactic clarity? If you had the rule in this state in the `:libunit_name` state, wouldn't it work the same way?

Reply:
I'm glad you asked ;-)

By using two ping-pong states like this, we require a strictly alternating pattern of ID and punctuation like `ID . ID , ID`, and we `pop!` as soon as the text deviates from that pattern. A single state would allow any sequence of ID and punctuation, including things like `ID ID . . ,`. This is a superset of the alternating patterns, so most of the time that would be fine.

The problem is that the `with` keyword in Ada is used for different things in different contexts. At the top-level outside any nested constructs, it is used a lot like the `import` keyword in Python or Ruby's `require`, and that is the case we're handling here. However, inside nested constructs it is used for other things. Take for example an extended version of the generic sort function in the visual sample:

```ada
generic
   type Element_Type is private;
   type Array_Type is array (Integer range <>) of Element_Type;
   with function ">" (Left, Right : Element_Type) return Boolean is <>;
   with function Random_Number return Integer is <>;
procedure Algorithms.Random_Sort (A : in out Array_Type);
```

In this use, `with` is followed by 5 identifier-like words that have nothing to do with namespaces and imports. If we got into state `:libunit_name` by mistake, we would incorrectly color all of `function Random_Number return Integer is` as a namespace.

We only enter `:libunit_name` following a `with` at the beginning of a line. This is how it is normally written, but as you can see from the example above, we're only three spaces away from incorrectly entering the state. I used two states so we would detect something like this earlier and revert to the `:root` state ASAP if something is wrong.

Another option would be to detect any keywords appearing in `:libunit_name` using the `self.class.idents` map. Keywords are not allowed as parts of library unit names, so we could bail on that too.